.github/workflows/tests.yml

```diff
@@ -31,29 +31,43 @@ jobs:
           - ubuntu-latest
           - macos-latest
           - windows-latest
-        python-version:
-          - "3.7"
-          - "3.8"
-          - "3.9"
-          - "3.10"
-          - "3.11"
-          - "pypy-3.7"
 
     steps:
       - name: "Check out the repo"
         uses: "actions/checkout@v3"
 
       - name: "Set up Python"
+        id: "setup-python"
         uses: "actions/setup-python@v4"
         with:
-          python-version: "${{ matrix.python-version }}"
-          cache: pip
-          cache-dependency-path: 'requirements/*.txt'
+          # The last listed Python version is the default.
+          python-version: |
+            pypy-3.7
+            3.7
+            3.8
+            3.9
+            3.10
+            3.11
+
+      - name: "Restore cache"
+        id: "restore-cache"
+        uses: "actions/cache@v3"
+        with:
+          path: |
+            .tox/
+            .venv/
+          key: "cache-python-${{ steps.setup-python.outputs.python-version }}-os-${{ runner.os }}-hash-${{ hashFiles('tox.ini', 'requirements/*.txt') }}"
```

Comment:
By caching the `.tox/` and `.venv/` directories, the virtual environment and the Tox environments can be restored as filesystem operations, rather than as an installation of Python packages.

The cache key includes a hash of `tox.ini` and all of the `requirements/*.txt` files changes. This ensures that a cache miss occurs if any of those files changes.

Note that it may be helpful to modify the cache key in the future. In particular, if the last-listed Python version above ("3.11" currently) doesn't change versions _but any of the other versions updates_ (say, from "3.10.x" to "3.10.y"), then:

1. This would result in a cache hit.
2. However, Tox would detect that the 3.10 version had changed and would rebuild the `py310` test environment.
3. The cache would not be updated with this new environment, resulting in an increase in test times.

This situation can be improved adding a step that runs each Python version and writes its exact version to a file. Then, this file could be included in the list of files passed to `hashFiles()`.

I haven't learned how to do this yet, so this improvement isn't included. The only thing that will happen in this situation is that Tox has to discard the Python 3.10 test environment and rebuild it, so test times will increase until the cache is invalidated by some other means, like when the `requirements/*.txt` files are updated.